source/h1_encoder.c

```diff
@@ -443,6 +497,36 @@ static void s_populate_chunk_line_buffer(
     AWS_ASSERT(wrote_chunk_line);
 }
 
+struct aws_h1_trailer *aws_h1_trailer_new(
+    struct aws_allocator *allocator,
+    const struct aws_http_headers *trailing_headers) {
+    /* Allocate trailer along with storage for the trailer-line */
+    size_t trailer_size = 0;
+    if (s_headers_size(trailing_headers, &trailer_size)) {
+        return NULL;
+    }
+
+    struct aws_h1_trailer *trailer = aws_mem_calloc(allocator, 1, sizeof(struct aws_h1_trailer));
+    trailer->allocator = allocator;
+    if (!trailer) {
+        return NULL;
+    }
+
+    if (aws_byte_buf_init(&trailer->trailer_data, allocator, trailer_size)) {
+        aws_mem_release(allocator, trailer);
+        return NULL;
+    }
+    s_write_headers(&trailer->trailer_data, trailing_headers);
+    return trailer;
+}
+
+void aws_h1_trailer_destroy(struct aws_h1_trailer *trailer) {
+    /* should this check be here? it makes it so that I need to check before calling this function */
+    AWS_PRECONDITION(trailer);
```

Comment:
for destroy() and release() calls, we usually add a check like:
```
if (trailer == NULL) {
    return;
}
```
so that they can be called blindly without checking if the ptr is null or not

so like, if destroy() worked that way, you could alter trailer_new() to do all its error handling like:
```
    if (step 1 goes wrong) {
        goto error;
    }

    if (step 2 goes wrong) {
        goto error;
    }

    if (step 3 goes wrong) {
        goto error;
    }

    ...
    /* success! */
    return trailer;
    
error:
    aws_h1_trailer_destroy(trailer);
    return NULL;
}
```